checks/src/main/kotlin/com/faithlife/lint/ForEachFunctionDetector.kt

```diff
@@ -10,23 +10,48 @@ import com.android.tools.lint.detector.api.Scope
 import com.android.tools.lint.detector.api.Severity
 import com.android.tools.lint.detector.api.SourceCodeScanner
 import com.intellij.psi.PsiMethod
+import org.jetbrains.kotlin.psi.KtLambdaExpression
 import org.jetbrains.uast.UCallExpression
 import org.jetbrains.uast.kotlin.KotlinUSafeQualifiedExpression
 
 class ForEachFunctionDetector : Detector(), SourceCodeScanner {
-    override fun getApplicableMethodNames(): List<String> = listOf("forEach", "forEachIndexed")
+    override fun getApplicableMethodNames(): List<String> =
+        listOf(FOR_EACH_NAME, FOR_EACH_INDEXED_NAME)
 
     override fun visitMethodCall(context: JavaContext, node: UCallExpression, method: PsiMethod) {
         if (node.uastParent is KotlinUSafeQualifiedExpression) return
 
+        val receiver = node.receiver
         val receiverTypeClass = context.evaluator.getTypeClass(node.receiverType)
 
-        if (receiverTypeClass != null && context.evaluator.inheritsFrom(receiverTypeClass, "java.lang.Iterable", false)) {
+        if (receiver != null && receiverTypeClass != null && context.evaluator.inheritsFrom(receiverTypeClass, "java.lang.Iterable", false)) {
+            val lambda = node.valueArguments.first().sourcePsi as KtLambdaExpression
+            val receiverName = receiver.asSourceString()
+
+            val replacementText = if (method.name == FOR_EACH_INDEXED_NAME) {
+                val firstParameter = lambda.valueParameters[0].text
+                val secondParameter = lambda.valueParameters[1].text
+                "for (($firstParameter, $secondParameter) in $receiverName.withIndex()) {"
+            } else {
+                val parameterName = lambda.valueParameters.firstOrNull()?.text ?: "it"
+                "for ($parameterName in $receiverName) {"
+            }
```

Comment:
A passed function reference will break this.

Reply:
Imo, `forEach*` with a function reference shouldn't trigger this incident. That case seems like it has some semantic benefit.

Reply:
I hadn't really considered differentiating along the syntax like that. The rationale would go back to either:

a) We think `forEach*` is bad because it's the only list operator that requires side effects to be useful.
b) We think `forEach*` is bad because break and continue are weird.

I tend to affirm both of those views, so I would say there's no real difference between a function reference and a lambda since the difference is essentially syntactic.

The exception for the safe navigation operator as-is isn't super consistent with my affirmation of a, as there are other means of converting the nullable type to something non-null `?:`, `orEmpty`, `ifNull`, etc.

So either I think we should remove the exception for safe navigation and don't exempt function reference parameter cases _or_ permit most uses and only raise an issue if a `forEach` has a `return` in it.